### PR TITLE
Signature padding improvements

### LINKED ISSUE #2841: ECDSA producing unverifiable or incorrect signatures for some curves
```cpp
#include <botan/ecdsa.h>
#include <botan/pubkey.h>
#include <botan/system_rng.h>
#include <botan/hash.h>
#include <botan/ber_dec.h>

static void sign_verify(const std::string& curve) {
    ::Botan::System_RNG rng;
    const std::vector<uint8_t> msg{0xFF};
    std::unique_ptr<::Botan::PK_Signer> signer;
    ::Botan::EC_Group group(curve);
    const ::Botan::BigInt priv_bn("1");
    const auto priv = std::make_unique<::Botan::ECDSA_PrivateKey>(::Botan::ECDSA_PrivateKey(rng, group, priv_bn));
    signer.reset(new ::Botan::PK_Signer(*priv, rng, "EMSA1(SHA-256)", ::Botan::DER_SEQUENCE));
    const auto signature = signer->sign_message(msg, rng);
    ::Botan::BER_Decoder decoder(signature);
    ::Botan::BER_Decoder ber_sig = decoder.start_sequence();

    size_t count = 0;

    ::Botan::BigInt R;
    ::Botan::BigInt S;
    while(ber_sig.more_items())
    {
        switch ( count ) {
            case    0:
                ber_sig.decode(R);
                break;
            case    1:
                ber_sig.decode(S);
                break;
            default:
                printf("Error: Too many parts in signature BER\n");
                abort();
        }

        ++count;
    }

    const auto pub_x = priv->public_point().get_affine_x();
    const auto pub_y = priv->public_point().get_affine_y();

    const ::Botan::PointGFp public_point = group.point(pub_x, pub_y);
    const auto pub = std::make_unique<::Botan::ECDSA_PublicKey>(::Botan::ECDSA_PublicKey(group, public_point));
    const auto sig = ::Botan::BigInt::encode_fixed_length_int_pair(R, S, group.get_order_bytes());
    auto hash = ::Botan::HashFunction::create("SHA-256");
    hash->update(msg);
    const auto _msg = hash->final();
    printf("%s; verified: %d\n", curve.c_str(), ::Botan::PK_Verifier(*pub, "Raw").verify_message(_msg, sig));

    return;
}
int main(void)
{
    const std::vector<std::string> curves{
        "brainpool160r1",
            "brainpool192r1",
            "brainpool224r1",
            "brainpool256r1",
            "brainpool320r1",
            "brainpool384r1",
            "brainpool512r1",
            "frp256v1",
            "gost_256A",
            "gost_512A",
            "secp160k1",
            "secp160r1",
            "secp160r2",
            "secp192k1",
            "secp192r1",
            "secp224k1",
            "secp224r1",
            "secp256k1",
            "secp256r1",
            "secp384r1",
            "secp521r1",
            "sm2p256v1",
            "x962_p192v2",
            "x962_p192v3",
            "x962_p239v1",
            "x962_p239v2",
            "x962_p239v3",
    };

    for (const auto& c : curves) {
        sign_verify(c);
    }
}
```

prints:

```
brainpool160r1; verified: 1
brainpool192r1; verified: 1
brainpool224r1; verified: 1
brainpool256r1; verified: 1
brainpool320r1; verified: 1
brainpool384r1; verified: 1
brainpool512r1; verified: 1
frp256v1; verified: 1
gost_256A; verified: 1
gost_512A; verified: 1
secp160k1; verified: 0
secp160r1; verified: 0
secp160r2; verified: 0
secp192k1; verified: 1
secp192r1; verified: 1
secp224k1; verified: 0
secp224r1; verified: 1
secp256k1; verified: 1
secp256r1; verified: 1
secp384r1; verified: 1
secp521r1; verified: 1
sm2p256v1; verified: 1
x962_p192v2; verified: 1
x962_p192v3; verified: 1
x962_p239v1; verified: 0
x962_p239v2; verified: 0
x962_p239v3; verified: 0
```

### PR DESCRIPTION
**Pull request by randombit**
Fixes #2841 and #3313

@reneme review is optional but I figure you would be interested. It is probably easier to review commit-by-commit. Also important note: Dilithium X.509 is still broken. The old tests used `Botan::get_sig_paddings` to get the padding parameters, but this returned an empty list for Dilithium. If I enable the tests for Dilithium, things fail.

#3319 captures some followup items.

### CODE REVIEW

File: src/lib/tls/tls_signature_scheme.cpp [hunk 1]
```diff
@@ -242,10 +242,40 @@ AlgorithmIdentifier Signature_Scheme::key_algorithm_identifier() const noexcept
...
+         return AlgorithmIdentifier(OID::from_string("RSA/EMSA4"),
+                                    hex_decode("3039A00F300D06096086480165030402020500A11C301A06092A864886F70D010108300D06096086480165030402020500A203020130A303020101"));
+      case RSA_PSS_SHA512:
+         return AlgorithmIdentifier(OID::from_string("RSA/EMSA4"),
+                                    hex_decode("3039A00F300D06096086480165030402030500A11C301A06092A864886F70D010108300D06096086480165030402030500A203020140A303020101"));
+
+      default:
+         // Note that Ed25519 and Ed448 end up here
+         return AlgorithmIdentifier();
+      }
```
Comment (reneme) [inline src/lib/tls/tls_signature_scheme.cpp:278]: Same as [last year](https://github.com/randombit/botan/pull/3125#discussion_r1049941524): Sad that we need to scatter those mappings around the code. But then again: it feels like every protocol has their own particularities that make it hard to abstract this.

Anyway: Maybe at least add a comment to the `hex_decode`-ed EMSA4 parameters explaining what's inside the magic strings?
In the linked comment, I regenerated those for every invocation. Less efficient but easier to understand:

```C++
   auto pss_params = [](const std::string hash_name)
      {
         const AlgorithmIdentifier hash_id(hash_name, AlgorithmIdentifier::USE_NULL_PARAM);
         const AlgorithmIdentifier mgf_id("MGF1", hash_id.BER_encode());

         std::vector<uint8_t> parameters;
         DER_Encoder(parameters)
            .start_sequence()
            .start_context_specific(0).encode(hash_id).end_cons()
            .start_context_specific(1).encode(mgf_id).end_cons()
            .start_context_specific(2).encode(HashFunction::create(hash_name)->output_length()).end_cons()
            .start_context_specific(3).encode(size_t(1)).end_cons() // trailer field
            .end_cons();

         return parameters;
      };
```

Still wondering whether `DER_Encoder` could do this as a `constexpr` generation. Purely academic question though.

**Review by reneme — COMMENTED** [review]
I skimmed over the changes and didn't find any obvious blunder. But frankly I'm lacking a good enough overview of the EMSA implementations to give a decent review.

Anything we can do to help with Dilithium, though?

File: src/lib/pk_pad/raw_hash/raw_hash.cpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+   {
+   m_bits += std::make_pair(input, length);
+   }
+
+void RawHashFunction::final_result(uint8_t out[])
+   {
+   if(m_output_length > 0 && m_bits.size() != m_output_length)
+      {
+      m_bits.resize(0);
+      throw Invalid_Argument("EMSA_Raw was configured to use a " +
```
Comment (lieser) [inline src/lib/pk_pad/raw_hash/raw_hash.cpp:22]: I guess this is a copy and paste error and `EMSA_Raw` should properly be `RawHashFunction`

Comment (randombit) [reply src/lib/pk_pad/raw_hash/raw_hash.cpp:22]: Ah thank you, yes

**Pull request merged**